haskell/cabal
# Haddock recompilation avoidance

Haddock no longer writes compilation files by default, so we do not need to pass tmp dirs for `-hidir`, `-stubdir`, and `-odir` via `--optghc`. Indeed, we do not *want* to do so, since it results in recompilation for every invocation of Haddock via Cabal. This commit stops this from happening for Haddock versions 2.28 or greater (when Hi Haddock was introduced).

This commit also prevents the definition of the `__HADDOCK_VERSION__` macro when invoking GHC through Haddock by default, which necessary to avoid recompilation during documentation generation, since the macro invalidates existing build results. Since a limited set of users may rely on this macro still, we introduce the `--haddock-version-cpp` flag and `haddock-version-cpp:` cabal.project field, which enable the definition of the `__HADDOCK_VERSION__` macro when invoking GHC through Haddock.

This commit also renames the `--haddock-lib` flag to `--haddock-resources-dir` (and `haddock-lib:` cabal.project field to `haddock-resources-dir:`), and adds this flag to the users guide since it was missing an entry. This also allows us to add this field to `cabal-install:test:integration-tests2`, since it is no longer ambiguous with the `--lib` flag.

This commit also causes `documentation: true` or `--enable-documentation` to imply `-haddock` for GHC.

Also, since Haddock >= 2.29 is renaming `--lib` to `--resources-dir`, this commit switches the flag provided to Haddock using a backwards compatible condition based on the Haddock version.

---
## QA Notes

Consider a Haskell package with dependencies. We should be able to simply configure the package so that doing `cabal build` followed by `cabal haddock` does not trigger a recompilation. Use the following `deps.cabal` file:
```cabal
cabal-version: 3.6
name: deps
version: 0.1.0

library
  build-depends:
      base
    , acme-missiles
  exposed-modules: Lib
```

The following `cabal.project` file:
```cabal
packages: .

package *
  documentation: True
  haddock-options: --optghc=-ddump-hi-diffs
```

And any `Lib.hs` will do:
```haskell
-- | Lib reexports 'launchMissiles'
module Lib (myMissiles, launchMissiles) where

import Acme.Missiles

-- | Another 'launchMissiles'
myMissiles :: IO ()
myMissiles = launchMissiles
```

This configuration should avoid recompilation during documentation generation. To verify this, with a package such as above, just run `cabal build`. Since documentation is enabled, the package will first build then generate Haddocks. Since Haddock is passing `-ddump-hi-diffs` to GHC, you should see output indicating that no recompilation checks have triggered during the Haddock phase:
```
Module flags unchanged
Optimisation flags changed; ignoring
HPC flags unchanged
signatures to merge in unchanged
[]
implementing module unchanged
Checking interface for module Prelude base
Module fingerprint unchanged
Checking interface for module Acme.Missiles cm-mssls-0.3-9160cec8
Module fingerprint unchanged
Checking interface for module GHC.Types ghc-prim
Module fingerprint unchanged
```

This means recompilation was successfully avoided.

**NOTE:** I am working on making `no-tmp-comp-dir` the default behavior for Haddock 2.29, but until then it is necessary to avoid recompilation.

---


Fixes  #9175
